### PR TITLE
feat(BB): PrepareProposal Chain

### PR DESCRIPTION
**Pull request by davidterpay**

### CODE REVIEW

**Comment by MaghnusM** [comment]
@davidterpay for me to come in and help here, can you start adding medium-detail comments to PRs?

File: blockbuster/abci/abci.go [hunk 1]
```diff
@@ -18,70 +15,36 @@ type (
...
-					h.logger.Error("failed to prepare lane; skipping", "lane", l.Name(), "err", err)
-					continue
-				}
-
-				for _, txBz := range laneTxs {
-					totalTxBytes += int64(len(txBz))
-
-					txHash := sha256.Sum256(txBz)
-					txHashStr := hex.EncodeToString(txHash[:])
+		proposal := h.prepareLanesHandler(ctx, blockbuster.Proposal{
```
Comment (alexanderbez) [inline blockbuster/abci/abci.go:41]: Is the idea that `prepareLanesHandler` modifies the `blockbuster.Proposal` argument passed in? If so, let's make sure we pass in a pointer.

File: blockbuster/lanes/auction/abci.go [hunk 2]
```diff
@@ -12,14 +12,19 @@ import (
...
-	bidTxIterator := l.Select(ctx, nil)
+func (l *TOBLane) PrepareLane(ctx sdk.Context, proposal blockbuster.Proposal, next blockbuster.PrepareLanesHandler) blockbuster.Proposal {
+	// Define all of the info we need to select transactions for the partial proposal.
+	txs := make([][]byte, 0)
 	txsToRemove := make(map[sdk.Tx]struct{}, 0)
 
+	// Calculate the max tx bytes for the lane and track the total size of the
+	// transactions we have selected so far.
+	maxTxBytes := blockbuster.GetMaxTxBytesForLane(proposal, l.cfg.MaxBlockSpace)
+	totalSize := int64(0)
```
Comment (alexanderbez) [inline blockbuster/lanes/auction/abci.go:23]:
```suggestion
	var int64 totalSize
```

File: blockbuster/utils.go [hunk 3]
```diff
@@ -32,3 +32,38 @@ func RemoveTxsFromLane(txs map[sdk.Tx]struct{}, mempool sdkmempool.Mempool) erro
...
+
+		return remainder
+	}
+
+	// Otherwise, we calculate the max tx bytes for the lane based on the ratio.
+	return ratio.MulInt64(proposal.MaxTxBytes).TruncateInt().Int64()
+}
+
+// UpdateProposal updates the proposal with the given transactions and total size.
+func UpdateProposal(proposal Proposal, txs [][]byte, totalSize int64) Proposal {
```
Comment (alexanderbez) [inline blockbuster/utils.go:57]: We should define methods on a `*Proposal` instead I think.

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**